File: algo/129. Sum Root to Leaf Numbers.py
# divide and conquer, too slow

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        if root is None:
            return 0
        return sum([int(item) for item in self.helper(root)])
        
    def helper(self, root: TreeNode) -> list():
        if root is None:
            return
        if root.left is None and root.right is None:
            return [str(root.val)]
        leftList = self.helper(root.left)
        rightList = self.helper(root.right)
        thisList = []
        if leftList: 
            thisList += [f'{root.val}{item}' for item in leftList]
        if rightList:
            thisList += [f'{root.val}{item}' for item in rightList]
        return thisList
        

File: round3/129. Sum Root to Leaf Numbers.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int SumNumbers(TreeNode root) {
        int res = 0;
        Helper(root, 0, ref res);
        return res;
    }
    
    private void Helper(TreeNode root, int currRes, ref int res) {
        if (root == null) {
            return;
        }
        
        int curr = currRes * 10 + root.val;        
        if (root.left == null && root.right == null) {
            res += curr;
        }        
        
        Helper(root.left, curr, ref res);
        Helper(root.right, curr, ref res);
    }
}